flink-tests/src/test/java/org/apache/flink/test/checkpointing/ZooKeeperHighAvailabilityITCase.java

```diff
@@ -0,0 +1,333 @@
+	 *     <li>Move away the contents of the ZooKeeper HA directory to make restoring from
+	 *       checkpoints impossible
+	 *     <li>Unblock the special function, which now induces a failure
+	 *     <li>Make sure that the job does not recover successfully
+	 *     <li>Move back the HA directory
+	 *     <li>Make sure that the job recovers, we use a latch to ensure that the operator
+	 *       restored successfully
+	 * </ol>
+	 */
+	@Test(timeout = 120_000L)
+	public void testRestoreBehaviourWithFaultyStateHandles() throws Exception {
+		CheckpointBlockingFunction.allowedInitializeCallsWithoutRestore.set(1);
+		CheckpointBlockingFunction.successfulRestores.set(0);
+		CheckpointBlockingFunction.illegalRestores.set(0);
+		CheckpointBlockingFunction.afterMessWithZooKeeper.set(false);
+		CheckpointBlockingFunction.failedAlready.set(false);
+
+		waitForCheckpointLatch = new OneShotLatch();
+		failInCheckpointLatch = new OneShotLatch();
+		successfulRestoreLatch = new OneShotLatch();
+
+		ClusterClient<?> clusterClient = miniClusterResource.getClusterClient();
+		final Deadline deadline = Deadline.now().plus(TEST_TIMEOUT);
+
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+		env.setParallelism(1);
+		env.setRestartStrategy(RestartStrategies.fixedDelayRestart(Integer.MAX_VALUE, 0));
+		env.enableCheckpointing(10); // Flink doesn't allow lower than 10 ms
+
+		File checkpointLocation = temporaryFolder.newFolder();
+		env.setStateBackend((StateBackend) new FsStateBackend(checkpointLocation.toURI()));
+
+		DataStreamSource<String> source = env.addSource(new UnboundedSource());
+
+		source
+			.keyBy((str) -> str)
+			.map(new CheckpointBlockingFunction());
+
+		JobGraph jobGraph = env.getStreamGraph().getJobGraph();
+		JobID jobID = Preconditions.checkNotNull(jobGraph.getJobID());
+
+		clusterClient.setDetached(true);
+		clusterClient.submitJob(jobGraph, ZooKeeperHighAvailabilityITCase.class.getClassLoader());
+
+		// wait until we did some checkpoints
+		waitForCheckpointLatch.await();
+
+		// mess with the HA directory so that the job cannot restore
+		File movedCheckpointLocation = temporaryFolder.newFolder();
+		int numCheckpoints = 0;
+		File[] files = haStorageDir.listFiles();
+		assertNotNull(files);
+		for (File file : files) {
+			if (file.getName().startsWith("completedCheckpoint")) {
+				assertTrue(file.renameTo(new File(movedCheckpointLocation, file.getName())));
+				numCheckpoints++;
+			}
+		}
+		// Note to future developers: This will break when we change Flink to not put the
+		// checkpoint metadata into the HA directory but instead rely on the fact that the
+		// actual checkpoint directory on DFS contains the checkpoint metadata. In this case,
+		// ZooKeeper will only contain a "handle" (read: String) that points to the metadata
+		// in DFS. The likely solution will be that we have to go directly to ZooKeeper, find
+		// out where the checkpoint is stored and mess with that.
```

Comment:
The solution would be to rename the checkpoint directory directly.

Reply:
You mean for now or for the future? I did that in an initial version but I wanted to verify that there are actually checkpoints that I'm moving and also there is other stuff in the HA directory that I don't want to move to not interfere with the test.